pkg/mutation/system.go

```diff
@@ -82,22 +84,28 @@ func (s *System) Upsert(m types.Mutator) error {
 func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) (bool, error) {
 	s.mux.RLock()
 	defer s.mux.RUnlock()
+	mutationUUID := uuid.New()
 	original := obj.DeepCopy()
 	maxIterations := len(s.orderedMutators) + 1
 
-	allAppliedMutations := [][]types.Mutator{}
+	var allAppliedMutations [][]types.Mutator
+	if *MutationLoggingEnabled || *MutationAnnotationsEnabled {
+		if allAppliedMutations == nil {
+			allAppliedMutations = [][]types.Mutator{}
+		}
+	}
 
 	for i := 0; i < maxIterations; i++ {
 		appliedMutations := []types.Mutator{}
 		old := obj.DeepCopy()
 		for _, m := range s.orderedMutators {
 			if m.Matches(obj, ns) {
-				mutated, err := m.Mutate(obj)
-				if mutated && *MutationLoggingEnabled {
+				mutated, err := m.Mutate(obj, mutationUUID)
```

Comment:
Instead of passing the mutation UUID all the way through, could we just prefix (or suffix) the error with the mutation UUID? That would avoid the need for downstream dependencies (like the mutate function) to be aware of it.

Reply:
Done